File: app/ai_request_status.py
"""Last completed investigation outcome; local metadata, never a provider probe."""
from datetime import datetime, timezone
from hashlib import sha256
import json
import os
from pathlib import Path
import tempfile
from threading import RLock
from typing import Literal

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, ValidationError, model_validator

STATUS_PATH = Path(__file__).resolve().parents[1] / 'data/local/ai-request-status.json'
_lock = RLock()
FailureKind = Literal['daily_quota', 'minute_quota', 'quota_unavailable', 'quota_unknown',
                      'configuration_missing', 'configuration_invalid', 'authentication',
                      'service_unavailable', 'network', 'timeout', 'invalid_response', 'analysis_incomplete']


class QuotaLimit(BaseModel):
    model_config = ConfigDict(extra='ignore')
    window: Literal['day', 'minute']
    measure: Literal['requests', 'tokens']
    limit: int | None = Field(default=None, ge=0, le=10**12, strict=True)


class Failure(BaseModel):
    model_config = ConfigDict(extra='ignore')
    kind: FailureKind
    limits: list[QuotaLimit] = Field(default_factory=list, max_length=20)


class Attempt(BaseModel):
    model_config = ConfigDict(extra='forbid')
    schema_version: Literal[1] = 1
    provider: Literal['gemini', 'ollama_local']
    context_id: str | None = Field(default=None, pattern=r'^[0-9a-f]{64}$')
    finished_at: AwareDatetime
    outcome: Literal['report_saved', 'report_unsaved', 'failed']
    failure: Failure | None = None
    source: Literal['application', 'prior_verification'] = 'application'

    @model_validator(mode='after')
    def consistent(self):
        if (self.outcome == 'failed') != (self.failure is not None):
            raise ValueError('Failure category must match outcome')
        if (self.source == 'application') != (self.context_id is not None):
            raise ValueError('Historical verification must not assert configuration identity')
        return self


def now_utc():
    return datetime.now(timezone.utc)


def capture_context() -> dict:
    from app.local_assistant import assistant_runtime
    runtime = assistant_runtime()
    values = [runtime['provider'], runtime['model'], os.getenv('GEMINI_BASE_URL', ''),
              os.getenv('GEMINI_API_KEY', ''), os.getenv('OLLAMA_BASE_URL', '')]
    return {'provider': runtime['provider'], 'model': runtime['model'],
            'configured': runtime['cloud_credentials_configured'] if runtime['provider'] == 'gemini' else True,
            'context_id': sha256(json.dumps(values, ensure_ascii=False).encode()).hexdigest()}


def _read(now):
    try:
        if STATUS_PATH.stat().st_size > 8192:
            return None, 'invalid'
        record = Attempt.model_validate_json(STATUS_PATH.read_text(encoding='utf-8'))
        if record.finished_at > now:
            return None, 'invalid'
        return record, 'available'
    except FileNotFoundError:
        return None, 'missing'
    except (ValidationError, ValueError, UnicodeError):
        return None, 'invalid'
    except OSError:
        return None, 'unreadable'


def _public(record, context, now, state='available'):
    result = {'provider': context['provider'], 'model': context['model'], 'configured': context['configured'],
              'is_live_check': False, 'current_availability': 'not_verified',
              'record_state': state, 'last_attempt': None}
    if record is None:
        return result
    if record.provider != context['provider'] or (record.context_id and record.context_id != context['context_id']):
        result['record_state'] = 'other_configuration'
        return result
    result['last_attempt'] = {'finished_at': record.finished_at.isoformat(), 'outcome': record.outcome,
        'failure': record.failure.model_dump(exclude_none=True) if record.failure else None,
        'source': record.source, 'configuration_match': record.context_id == context['context_id'],
        'older_than_24h': (now - record.finished_at).total_seconds() > 86400}
    return result


def request_status(context: dict | None = None) -> dict:
    context = context or capture_context()
    now = now_utc()
    with _lock:
        record, state = _read(now)
    return _public(record, context, now, state)


def _failure(error):
    kind = getattr(error, 'kind', 'analysis_incomplete')
    details = getattr(error, 'provider_error', None)
    limits = []
    if isinstance(details, dict) and details.get('kind') in {'daily_quota', 'minute_quota', 'quota_unavailable', 'quota_unknown'}:
        kind = details['kind']
        supplied = details.get('limits', [])
        if isinstance(supplied, list):
            for item in supplied[:20]:
                try:
                    limits.append(QuotaLimit.model_validate(item))
                except (ValidationError, TypeError):
                    continue
    try:
        return Failure(kind=kind, limits=limits)
    except ValidationError:
        return Failure(kind='analysis_incomplete')


def record_outcome(context: dict, outcome: str, *, error=None) -> dict:
    if context['provider'] not in {'gemini', 'ollama_local'}:
        return {**_public(None, context, now_utc(), 'unsupported_configuration'), 'recording_saved': False}
    written = False
    temporary = None
    with _lock:
        now = now_utc()
        record = Attempt(provider=context['provider'], context_id=context['context_id'], finished_at=now,
                         outcome=outcome, failure=_failure(error) if outcome == 'failed' else None)
        try:
            STATUS_PATH.parent.mkdir(parents=True, exist_ok=True)
            with tempfile.NamedTemporaryFile(mode='w', encoding='utf-8', dir=STATUS_PATH.parent,
                                             prefix='.ai-status-', suffix='.tmp', delete=False) as handle:
                temporary = Path(handle.name)
                handle.write(record.model_dump_json())
            os.replace(temporary, STATUS_PATH)
            written = True
        except OSError:
            pass  # An optional status record must not turn a usable report into a failure.
        finally:
            if temporary:
                try:
                    temporary.unlink(missing_ok=True)
                except OSError:
                    pass
    return {**_public(record, context, now), 'recording_saved': written}

File: app/assistant_ui/ai-request-status.js
/* Public status copy is built from bounded categories, never raw upstream text. */
const AIRequestStatus = (()=>{
  const failureNames={daily_quota:'每日额度限制',minute_quota:'每分钟额度限制',quota_unavailable:'配额不可用',
    quota_unknown:'额度受限，类别未确认',configuration_missing:'密钥未配置',configuration_invalid:'请求配置或格式不匹配',
    authentication:'认证失败',service_unavailable:'服务暂不可用',network:'网络连接失败',timeout:'请求超时',
    insufficient_balance:'API 账户余额不足',rate_limit:'调用频率受限',
    invalid_response:'返回结果未通过校验',analysis_incomplete:'未得到完整报告'};
  function providerName(provider){
    const names={deepseek:'DeepSeek',gemini:'Gemini',ollama_local:'本地模型'};
    return Object.hasOwn(names,provider)?names[provider]:'AI';
  }
  function runtimeView(runtime){
    const cloud=runtime.inference_location==='cloud',name=providerName(runtime.provider);
    const model=typeof runtime.model==='string'?runtime.model:'模型未确认';
    const bounded=runtime.investigation_timeout_seconds===120 && runtime.transient_attempt_limit===3;
    return {
      label:cloud && runtime.cloud_credentials_configured===false
        ?`${name} 密钥未配置 · 设备资料可查`
        :`${cloud?name+' 云端推理':'本机推理'} · ${model}`,
      footer:`${cloud?name+' 云端推理':'本机推理'} · 只读辅助分析 · 报告保存在本机`,
      note:cloud?`分析时会将所选设备的证据和输入内容发送至 ${name}，报告保存在本机。`
        +(bounded?'单次排查最长等待 120 秒。':'')
        :'分析与模型推理运行于本机。'
    };
  }
  function view(data){
    const local='已载入的运行数据、故障资料和备件目录仍可查看。';
    if(!data)return {title:'请求状态暂时无法读取',tone:'warning',time:null,detail:'没有进行 AI 调用。'+local};
    if(data.configured===false)return {title:providerName(data.provider)+' 密钥未配置',tone:'warning',time:null,
      detail:'完整 AI 排查需要后端配置有效密钥。'+local};
    if(data.record_state==='unsupported_configuration')return {title:'AI 配置无法识别',tone:'warning',time:null,detail:local};
    const last=data.last_attempt;
    if(!last || !Number.isFinite(Date.parse(last.finished_at)))return {
      title:data.record_state==='other_configuration'?'配置已变化，尚无对应排查记录':'尚无可核验的完整排查记录',
      tone:'neutral',time:null,detail:'当前配置尚未完成分析验证。'};
    const prior=last.source==='prior_verification' || last.configuration_match!==true;
    const prefix=prior?'此前配置的记录：':'最近一次排查：';
    let title, tone='neutral';
    if(last.outcome==='report_saved')title=prefix+'报告已生成并保存';
    else if(last.outcome==='report_unsaved'){title=prefix+'报告已生成，保存失败';tone='warning';}
    else if(last.outcome==='failed'){title=prefix+(Object.hasOwn(failureNames,last.failure?.kind)?failureNames[last.failure.kind]:'未完成');tone='warning';}
    else return {title:'排查记录无法识别',tone:'warning',time:null,detail:local};
    let detail=prior?'对应当时配置；当前配置尚未核验。':'这是上次完整排查的结果，不代表当前可用性。';
    if(last.older_than_24h)detail+='记录已超过 24 小时。';
    if(['daily_quota','quota_unavailable'].includes(last.failure?.kind))detail+='请在服务商账户查看可用额度。';
    if(last.failure?.kind==='insufficient_balance')detail+='请核查 API 账户余额。';
    if(last.failure?.kind==='rate_limit')detail+='请降低请求频率后再手动尝试。';
    if(data.recording_saved===false)detail+='本次状态未写入本机记录，刷新后可能缺失。';
    if(last.outcome!=='report_saved')detail+=local;
    return {title,tone,time:last.finished_at,detail};
  }
  return {view,providerName,runtimeView};
})();
if(typeof module!=='undefined')module.exports=AIRequestStatus;
